File: Kth Largest Element.cpp
class Solution {
public:
    /*
     * param k : description of k
     * param nums : description of array and index 0 ~ n-1
     * return: description of return
     */
    int kthLargestElement(int k, vector<int> nums) {
        // write your code here
        ////////////////Time Limit Exceeded, without random////////////////
        // return quickSelect(nums.size() - k + 1, 0, nums.size() - 1, nums);
        ////////////copy the answer///////////////
        ///short one
        // nth_element(nums.begin(), nums.begin() + k - 1, nums.end(), greater<int>());
        // return nums[k - 1];
        ///quickselect with random
        int left = 0, right = nums.size() - 1;
        default_random_engine gen((random_device())());
        while (left <= right)
        {
            uniform_int_distribution<int> dis(left, right);
            int pivot = dis(gen);
            int new_pivot = partitionAroundPivot(left, right, pivot, nums);
            if (new_pivot > k - 1)
            {
                right = new_pivot - 1;
            }
            else if (new_pivot < k - 1)
            {
                left = new_pivot + 1;
            }
            else
            {
                return nums[new_pivot];
            }
            
        }
    }
    
    int partitionAroundPivot(int start, int end, int pivot, vector<int> &nums)
    {
        int pivot_val = nums[pivot];
        swap(nums[end], nums[pivot]);
        int new_pivot = start;
        for (int i = start; i < end; ++i)
        {
            if (nums[i] > pivot_val)
            {
                swap(nums[i], nums[new_pivot]);
                ++new_pivot;
            }
        }
        swap(nums[end], nums[new_pivot]);
        return new_pivot;
    }
    
    int quickSelect(int k, int start, int end, vector<int> &nums)
    {
        //quick sort
        int part = position(start, end, nums);
        if (part > k - 1)
        {
            return quickSelect(k, start, part - 1, nums);
        }
        else if (part < k - 1)
        {
            return quickSelect(k, part + 1, end, nums);
        }
        else
        {
            return nums[part];
        }
        
    }
    int position(int start, int end, vector<int> &nums)
    {

        int left = start;
        int right = end;
        int target = nums[left];
        while (left < right)
        {
            while (nums[right] > target && right > left)
            {
                --right;
            }
            if (right > left)
            {
                nums[left] = nums[right];   
            }
            while (nums[left] < target && right > left)
            {
                ++left;
            }
            if (right > left)
            {
                nums[right] = nums[left];
            }
        }
        nums[left] = target;
        return left;
    }
};

File: Kth Largest Element.py
import random
class Solution:
    # @param k & A a integer and an array
    # @return ans a integer
    def kthLargestElement(self, k, A):
        
        if not A:
            return
        left, right = 0, len(A) - 1
        while left <= right:
            pivot = random.randint(left, right)
            #change pivot
            A[right], A[pivot] = A[pivot], A[right]
            newpivot = left
            for i in range(left, right):
                if A[i] > A[right]:
                    A[newpivot], A[i] = A[i], A[newpivot]
                    newpivot += 1
                    
            A[newpivot], A[right] = A[right], A[newpivot]
            
            if newpivot + 1 == k:
                return A[newpivot]
            elif newpivot + 1 > k:
                right = newpivot - 1
            else:
                left = newpivot + 1
        # return A[left]


import random
class Solution:
    # @param k & A a integer and an array
    # @return ans a integer

    def kthLargestElement(self, k, A):
        def swap(index1, index2, A):
            A[index1], A[index2] = A[index2], A[index1]
           
        if not A:
            return None
            
        length = len(A)
        if length < k:
            return None
        
        left, right = 0, length - 1
        while left <= right:
            index = random.randint(left, right)
            swap(index, right, A)
            # target = 
            pivot = left
            for i in range(left, right):
                if A[i] > A[right]:
                    swap(i, pivot, A)
                    pivot += 1
            
            swap(right, pivot, A)
            if pivot + 1 == k:
                return A[pivot]
            elif pivot + 1 < k:
                left = pivot + 1
            else:
                right = pivot - 1
        
        return None



import random
class Solution:
    # @param k & A a integer and an array
    # @return ans a integer
    def kthLargestElement(self, k, A):
        ## [5, 5, 5, 5, 5, 5]
        
        if not A:
            return None
        length = len(A)
        if k > length or k < 1:
            return None
        left, right = 0, length - 1
        while left <= right:
            pivot = random.randint(left, right)
            ## three colors
            ## for testcase: [5, 5, 5, 5, 5, 5] with replication
            target = A[pivot]
            start, end = left, right
            pre = start
            while start <= end:
                if A[start] > target:
                    A[pre], A[start] = A[start], A[pre]
                    pre += 1
                    start += 1
                elif A[start] == target:
                    start += 1
                else:
                    A[end], A[start] = A[start], A[end]
                    end -= 1
            if pre + 1 <= k <= end + 1:
                return A[pre]
            elif pre + 1 > k:
                right = pre - 1
            else:
                left = end + 1
            
            # target = A[pivot]
            # A[right], A[pivot] = A[pivot], A[right]
            # pre = left
            # for i in range(left, right):
            #     if A[i] > target:
            #         A[pre], A[i] = A[i], A[pre]
            #         pre += 1
            # A[right], A[pre] = A[pre], A[right]
            # if pre + 1 == k:
            #     return A[pre]
            # elif pre + 1 > k:
            #     right = pre - 1
            # else:
            #     left = pre + 1
        # raise Exception
        return None